activesupport/lib/active_support/messages/metadata.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+require "time"
+
+module ActiveSupport
+  module Messages
+    class Metadata #:nodoc:
+      def initialize(expires_at, purpose)
+        @expires_at, @purpose = expires_at, purpose
+      end
+
+      class << self
+        def wrap(message, expires_at: nil, expires_in: nil, purpose: nil)
+          { "value" => message, "_rails" => { "exp" => pick_expiry(expires_at, expires_in), "pur" => purpose } }
+        end
+
+        def verify(message, purpose)
+          metadata = extract_metadata(message)
+          metadata.nil? ? message : (metadata.match?(purpose) && metadata.fresh?) ? message["value"] : nil
+        end
+
+        private
+
+          def pick_expiry(expires_at, expires_in)
+            expires_at ? expires_at.utc.iso8601(3) : (expires_in ? expires_in.from_now.utc.iso8601(3) : nil)
+          end
+
+          def extract_metadata(message)
+            if message_contains_metadata?(message)
```

Comment:
I don't think this extra abstraction is pulling its weight. When I read the next line I'm confused as to where `message["_rails"]` came from. I'd just put the method body here.